### ISSUE TITLE
Expression: autocomplete should be more helpful

### ISSUE THREAD

**Issue by kriswilk** [FreeCAD/FreeCAD#16848]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

**CASE STUDY**:

Creating an expression referring to a VarSet (or Spreadsheet) property of, eg. `hole_spacing`.

**DESIRED**:

Typing `h` should immediately find and prioritize `<<VarSet>>.hole_spacing`.

**CURRENTLY**:

First you must know to type/autocomplete `<<VarSet>>.`. Typing the dot `.` by hand will at least present a list of items. Comically, if you let autocomplete fill in `<<VarSet>>.` (with the dot), no list is presented until you start typing more.

**RATIONALE**:

Improving the intelligence of the autocompletion will:

- save time, keystrokes, and mental energy for all users.
- prevent new users from a needless an excursion to find help.

FWIW, I also think that (as in most other software) pressing TAB should complete the top entry from the autocomplete list.

### Full version info

```shell
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 1.0.0RC2.38806 (Git)
Build type: Release
Branch: (HEAD detached at 1.0rc2)
Hash: 3d63fc6c2f665a8d5e6468845a419bcac80756c7
Python 3.11.9, Qt 5.15.13, Coin 4.0.3, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: OpenLight.qss/OpenLight/Qt default
Installed mods: 
  * OpenTheme 2024.9.1
```

### Subproject(s) affected?

Expressions

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by hooperbloob** [comment]
see also  https://github.com/FreeCAD/FreeCAD/issues/15065

**Comment by maxwxyz** [comment]
Related: https://github.com/FreeCAD/FreeCAD/issues/5644

**Comment by maxwxyz** [comment]
Is this a duplicate of: https://github.com/FreeCAD/FreeCAD/issues/12796?

**Comment by kriswilk** [comment]
I will also add that, in general, the search should order results that _begin_ with the search text higher than those where it's merely a substring. For example here, when I type `c`, `Constraints` is much more likely to be what I want than any `Sketch`.

![Screenshot 2024-09-27 092833](https://github.com/user-attachments/assets/7fdd40c4-8ec3-450a-80a6-882adcd8e982)

**Comment by GAZ082** [comment]
These are really good observations. As well as other ones in separate issues (the TAB to trigger autocompletion is one i even mentioned in the Discord a few weeks back).

- Give priority to SpreadSheet/Varset.
  - How to deal if both are in a document?
- Rather than "containing" a character, should be a "starts with".
  - This may have some consequences as well worth exploring. But I do not forsee any negatives.

Can't we group tickets like these somehow under a "expression enhancement" project or something like that? I feel these issues can be solved in one go.

**Comment by kriswilk** [comment]
I agree that a few related issues could be collected under the general umbrella of improving the expression autocomplete experience. Perhaps the dev team will consolidate them in some way.

To clarify my "containing" vs. "starts with" suggestion: In the [example above](https://github.com/FreeCAD/FreeCAD/issues/16848#issuecomment-2379301971), items such as `Sketch` should NOT be eliminated... I only meant that objectively "better" matches like `Constraints` should be higher in the list. This is all basic "fuzzy searching" logic that, as you say @GAZ082, can probably be solved in one go.

As for which should come first between VarSet and Spreadsheet variables (if multiple, equally good matches exist)... typical fuzzy search would revert to alpha order. So `Spreadsheet.myvar` before `VarSet.myvar`.

I look forward to the day that I can just type `dia` and see a tidy list of all my custom diameter properties to pick from. 🙏

**Comment by pieterhijma** [comment]
Relevant to this PR is https://github.com/FreeCAD/FreeCAD/issues/13038. I actually have PR https://github.com/FreeCAD/FreeCAD/pull/13112 in draft that introduces "named scopes". The PR sits there because it's a large one and it is a bit unclear how to incorporate this kind of functionality, but I still like the feature to create a scope properties, name the scope, and use the scope for autocompletion, adding properties, etc. The main idea is that the scope is globally available to users.
